Add support for webhook API calls

chadworth opened the pull request.

chadworth commented:
@harlow would you be able to take a look at this PR? thanks!

harlow commented on src/main/java/com/clearbit/client/api/CompanyApi.java:
```diff
@@ -50,17 +50,47 @@ public Company stream(String domain) throws ApiException {
+     }
+     
+    // create path and map variables
+    String path = "https://company-stream.clearbit.com/v1/companies/domain/{domain}?webhook_id={webhookId}"
```
this should be using the non-streaming endpoint `https://company.clearbit.com`

chadworth replied:
good catch. done

harlow commented on src/main/java/com/clearbit/client/api/CompanyApi.java:
```diff
@@ -50,17 +50,47 @@ public Company stream(String domain) throws ApiException {
+    // create path and map variables
+    String path = "https://company-stream.clearbit.com/v1/companies/domain/{domain}?webhook_id={webhookId}"
+        .replaceAll("\\{" + "domain" + "\\}", apiClient.escapeString(domain.toString()))
+        .replaceAll("\\{" + "webhookId" + "\\}", apiClient.escapeString(webhookId.toString()));
```
Totally valid way to pass a webhook. You can also add one to your account at https://dashboard.clearbit.com/account in the "webhook URL" field

chadworth replied:
Those are different things, right? The webhook URL can only be added via your account dashboard. The webhook ID as I understand it is a way of matching up request and response and can be any arbitrary identifier.

harlow replied:
Yes- good catch! You are correct. Webhook ID is the identifier of the email/company in the callers system.

harlow commented on src/main/java/com/clearbit/client/api/CompanyApi.java:
```diff
@@ -29,16 +29,16 @@ public ApiClient getApiClient() {
-   * Person API
-   * The Person API lets you retrieve social information associated with an email address, such as a person’s name, location and Twitter handle.
-   * @param email the person’s email address
-   * @return Person
```
Ahh nice catch!

harlow commented on src/main/java/com/clearbit/client/api/CompanyApi.java:
```diff
@@ -29,16 +29,16 @@ public ApiClient getApiClient() {
-   * @param email the person’s email address
-   * @return Person
+   * Company API
+   * The Company API lets you lookup company data via a domain name.
```
can we add a note here that the `stream` is a blocking operation. i.e. if we don't have the Company in our database it will hold the connection open until the lookup has completed. Typically 3-5 seconds.

chadworth replied:
done

harlow commented on src/main/java/com/clearbit/client/api/CompanyApi.java:
```diff
@@ -50,17 +50,47 @@ public Company stream(String domain) throws ApiException {
+    
+     // verify the required parameters are set
+     if (domain == null) {
+        throw new ApiException(400, "Missing the required parameter 'domain' when calling CompanyApi.stream");
```
Still says `CompanyAPI.stream`

chadworth replied:
done

harlow commented on src/main/java/com/clearbit/client/api/CompanyApi.java:
```diff
@@ -50,17 +50,47 @@ public Company stream(String domain) throws ApiException {
+   * @param domain the company's website domain
+   * @return Company
+   */
+  public Company webhook(String domain, String webhookId) throws ApiException {
```
wondering about naming here. When we change the url back to `company.clearbit.com` this is just considered a `lookup` and the `webhookId` is an optional parameter (not required)

harlow replied:
maybe `Company lookup` and `Company streamingLookup` ?

chadworth replied:
done

harlow commented on src/main/java/com/clearbit/client/api/CompanyApi.java:
```diff
@@ -50,17 +50,47 @@ public Company stream(String domain) throws ApiException {
     TypeRef<Company> returnType = new TypeRef<Company>() {};
-    return apiClient.invokeAPI(path, "GET", queryParams, postBody, postBinaryBody, headerParams, formParams, accept, contentType, authNames, returnType);
+    Company company = apiClient.invokeAPI(path, "GET", queryParams, postBody, postBinaryBody, headerParams, formParams, accept, contentType, authNames, returnType);
+    return company == null || company.getId() == null ? null : company;
```
w/ the regular non-streaming API here is the req/rep pattern:

```
# Company exists in DB
1. Curl URL
2. Company JSON returned

# Company doesn't exist in DB
1. Curl URL
2. Lookup Enqueued 202 returned (domain put into lookup queues, webhooks automatically delivered if set)
# ... several seconds later 
3. Curl URL
4. Company JSON returned
```

It's the 202 case we may need to code against here

harlow commented on src/main/java/com/clearbit/client/api/PersonApi.java:
```diff
@@ -50,17 +50,47 @@ public Person stream(String email) throws ApiException {
+     }
+     
+    // create path and map variables
+    String path = "https://person.clearbit.com/v1/people/email/{email}?webhook_id={webhookId}"
```
Just an FYI in this case only Person information will be returned (no company record). We also have a `combined` endpoint which returns a payload like this:

```
{
  id: 'xyz',
  type: 'person_company',
  body: {
    person: {},
    company: {}
  }
}
```

chadworth commented:
I've addressed all the comments, so I'm going to go ahead and merge it. Please let me know if there are any other issues.

Looking forward to seeing this published in Maven!

harlow commented:
Looks great. Thanks for knocking out all the naming stuff. :+1:
